### PR TITLE
feat: add distribution schedules #LSC-222

### PR DESCRIPTION
**Pull request by zavgorodnii**
1. Compared to https://github.com/Anchor-Protocol/anchor-token-contracts/tree/main/contracts/staking, you can now add distribution periods;
2. All `CanonicalAddr`s have been replaced by just `Addr`s;
3. A `distribution_account` config is added that gets the rights to migrate the contract.

### CODE REVIEW

File: contracts/staking/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    }
+}
+
+pub fn add_distribution_periods(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    periods: Vec<(u64, u64, Uint128)>,
+) -> StdResult<Response> {
+    let mut config: Config = read_config(deps.storage)?;
```
Comment (swelf19) [inline contracts/staking/src/contract.rs:72]: unauthorized access?

File: contracts/staking/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    config.distribution_schedule.extend(periods);
```
Comment (swelf19) [inline contracts/staking/src/contract.rs:78]: distribution slots can overlap each other, is it ok?

File: contracts/staking/src/state.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+static KEY_STATE: &[u8] = b"state";
+
+static PREFIX_REWARD: &[u8] = b"reward";
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Config {
+    pub distribution_account: Addr,
+    pub ldo_token: Addr,
+    pub staking_token: Addr,
+    pub distribution_schedule: Vec<(u64, u64, Uint128)>,
```
Comment (swelf19) [inline contracts/staking/src/state.rs:17]: use struct with named fields instead of tuple?

File: contracts/staking/src/contract.rs (hunk 2 above)
Comment (zavgorodnii) [reply contracts/staking/src/contract.rs:78]: Yup, that's ok

File: contracts/staking/src/contract.rs (hunk 1 above)
Comment (zavgorodnii) [reply contracts/staking/src/contract.rs:72]: +++

File: contracts/staking/src/state.rs (hunk 3 above)
Comment (zavgorodnii) [reply contracts/staking/src/state.rs:17]: Wanted to change as little as possible

**Comment by zavgorodnii** [comment]
also there's no tests guys

File: contracts/staking/src/testing.rs [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+            staker: "addr0000".to_string(),
+            reward_index: Decimal::from_ratio(25000u64, 1u64),
+            pending_reward: Uint128::from(3000000u128),
+            bond_amount: Uint128::from(100u128),
+        }
+    );
+}
+
+#[test]
+fn test_withdraw() {
```
Comment (swelf19) [inline contracts/staking/src/testing.rs]: is it possible to make some tests with multiple stakers to test reward distribution?

Comment (zavgorodnii) [reply contracts/staking/src/testing.rs]: I guess so, will add this as an additional task.

Comment (mmikev) [reply contracts/staking/src/testing.rs]: Also would be nice to have tests for overlapping distribution slots

Comment (mmikev) [reply contracts/staking/src/testing.rs]: And integration tests

Comment (zavgorodnii) [reply contracts/staking/src/testing.rs]: O.K., will add more tests; integration tests are out of scope for this PR

**Review by swelf19 — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Pull request merged**